### DISCUSSION TITLE
SkiaSharp encode to byteArray result is not the same as System.Drawing.Image.Save

### DISCUSSION THREAD

**Discussion by WillisLiou** [mono/SkiaSharp#2570] [Q&A]
I am now trying to upgrade my project from .net framework 4.7.2 to .net Standard 2.0. Since my project also need to run in Linux, for image processing (ex. Bitmap), I choose SkiaSharp which was also recommended in MSDN.

My question now is that when I trying to save bmp as different format (ex. Png or Jpeg) and get byte array, byte array result is different.

Since compatibility issue, is it possible that the output byte array can be the same?

**Code used Syste.Drawing**

```cs
....
var encoder = GetEncoder(imageFormat);
if (encoder == null)
{
    return Array.Empty<byte>();
}

var myEncoder = Encoder.Quality;
var myEncoderParameters = new EncoderParameters(1);
var myEncoderParameter = new EncoderParameter(myEncoder, 50L);
myEncoderParameters.Param[0] = myEncoderParameter;

byte[] byteArray;
using (var stream = new MemoryStream())
{
    bmp.Save(stream, encoder, myEncoderParameters);
    byteArray = stream.ToArray();
}
....

public ImageCodecInfo GetEncoder(ImageFormat format)
{
    try
    {
        var imageDecoders = ImageCodecInfo.GetImageDecoders();

        return imageDecoders.FirstOrDefault(codec => codec.FormatID == format.Guid);
    }
    catch (Exception)
    {
        return null;
    }
}
```

**Code used SkiaSharp**
```cs
public byte[] BmpToByteArrayByImageFormat(SKBitmap bmp, SKEncodedImageFormat imageFormat)
{
    try
    {
        var data = bmp.Encode(imageFormat, 50);

        return data.ToArray();
    }
    catch (Exception)
    {
        return Array.Empty<byte>();
    }
}
```

**Comment by TommiGustafsson-HMP** [comment]
Have you checked, for example, with `pngcheck` and GIMP if the compression, bit depth, or color space parameters or EXIF and XMP metadata differ in the resulting PNG files?

**Reply by mattleibow** [reply]
Not only might there be different metadata, there may also be a slightly different encoding algorithm as System.Drawing encoding is likely using Win32 things, but SkiaSharp is using libpng.

If you are sure the bitmap is the same, it may be better to load/decode the image and compare pixel data as that will stay the same.
